### PR TITLE
add a currency entity for ether

### PR DESCRIPTION
**Pull request by moodysalem**

### CODE REVIEW

**Comment by moodysalem** [comment]
still wip since need to support ether as an input/output in routes/trades and potentially also in pairs (since v1 has ether pairs) at which point this sdk can be used for either v1 or v2 trades, which is what we already do in the UI

**Comment by moodysalem** [comment]
we don't support ether in pairs, only in trades and routes. this pr also adds a router class that represents the on-chain router and allows generating call parameters for a given trade

File: src/router.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  // whether any of the tokens in the path are fee on transfer tokens, which should be handled with special methods
+  feeOnTransfer?: boolean
+}
+
+export interface SwapParameters {
+  // the method to call on the Uniswap V2 Router
+  methodName: string
+  // the arguments to pass to the method, all hex encoded
+  args: (string | string[])[]
+}
```
Comment (NoahZinsmeister) [inline src/router.ts:25]: i think we should add value (`msg.value`) as a parameter

File: test/pair.test.ts [hunk 2]
```diff
@@ -1,12 +1,85 @@
...
+      )
+    })
+  })
+
   describe('#getAddress', () => {
     it('returns the correct address', () => {
-      const usdc = new Token(ChainId.MAINNET, '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48', 18, 'USDC', 'USD Coin')
-      const dai = new Token(ChainId.MAINNET, '0x6B175474E89094C44Da98b954EedeAC495271d0F', 18, 'DAI', 'DAI Stablecoin')
-      expect(Pair.getAddress(usdc, dai)).toEqual('0xAE461cA67B15dc8dc81CE7615e0320dA1A9aB8D5')
+      expect(Pair.getAddress(USDC, DAI)).toEqual('0xAE461cA67B15dc8dc81CE7615e0320dA1A9aB8D5')
```
Comment (NoahZinsmeister) [inline test/pair.test.ts:18]: oh i was thinking, this should probably return a token. maybe keep the method with intent to deprecate and add a `Pair.getLiquidityToken` or something?

Comment (NoahZinsmeister) [reply test/pair.test.ts:18]: maybe not worth though idk

File: test/router.test.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
+import invariant from 'tiny-invariant'
+import { ChainId, CurrencyAmount, ETHER, Pair, Percent, Route, Router, Token, TokenAmount, Trade, WETH } from '../src'
+import JSBI from 'jsbi'
+
+function checkDeadline(deadline: string[] | string): void {
+  expect(typeof deadline).toBe('string')
+  invariant(typeof deadline === 'string')
```
Comment (NoahZinsmeister) [inline test/router.test.ts:7]: don't need this

Comment (moodysalem) [reply test/router.test.ts:7]: toBe doesn't work as a type assertion but invariant is

Comment (NoahZinsmeister) [reply test/router.test.ts:7]: oh got it

File: test/pair.test.ts (hunk 2 above)
Comment (moodysalem) [reply test/pair.test.ts:18]: not sure, maybe consider in another pr. the way we use it in the interface in both cases, creating token instances would be a waste

**Pull request merged**